### ISSUE TITLE
Error while indexing. Centos 7 Postgresql 12

### ISSUE THREAD

**Issue by boooch** [osm-search/Nominatim#2045]
2020-11-04 19:38:20 == module path: /srv/nominatim/build/module
2020-11-04 19:38:20 == Index ranks 0 - 4
'/srv/nominatim/Nominatim-3.5.2/nominatim/nominatim.py' --database nominatim --port 5432 --threads 15 -v --maxrank 4WARNING: Starting indexing rank (0 to 4) using 15 threads
WARNING: Starting rank 0
WARNING: Done 0/0 in 0 @ 0.000 per second - FINISHED rank 0

WARNING: Starting rank 1
WARNING: Done 0/0 in 0 @ 0.000 per second - FINISHED rank 1

WARNING: Starting rank 2
WARNING: Done 0/0 in 0 @ 0.000 per second - FINISHED rank 2

WARNING: Starting rank 3
WARNING: Done 0/0 in 0 @ 0.000 per second - FINISHED rank 3

WARNING: Starting rank 4
INFO: Done 100 in 0 @ 35599.858 per second - rank 4 ETA (seconds): 0.00
Traceback (most recent call last):
  File "/srv/nominatim/Nominatim-3.5.2/nominatim/nominatim.py", line 326, in <module>
    Indexer(options).run()
  File "/srv/nominatim/Nominatim-3.5.2/nominatim/nominatim.py", line 206, in run
    self.index(RankRunner(self.maxrank), 20)
  File "/srv/nominatim/Nominatim-3.5.2/nominatim/nominatim.py", line 243, in index
    t.wait()
  File "/srv/nominatim/Nominatim-3.5.2/nominatim/nominatim.py", line 133, in wait
    wait_select(self.conn)
  File "/usr/local/lib64/python3.6/site-packages/psycopg2/extras.py", line 775, in wait_select
    state = conn.poll()
psycopg2.errors.TriggeredDataChangeViolation:  ERROR: tuple to be updated was already modified by an operation triggered by the current command
HINT: Consider using an AFTER trigger instead of a BEFORE trigger to propagate changes to other rows.

**Comment by boooch** [comment]
this query triggered error: 
UPDATE placex SET indexed_status = 0 WHERE place_id IN (123684,14854538,128578,14714287,311084,14778481,14844284,14794397,14496415,14389638,14445874,128021,14833047,14413001,110793,14713753,110580,14827990,123217,14778210)

**Comment by woodpeck** [comment]
Could you post the result of 
```
    SELECT osm_id,osm_type 
    FROM placex 
    WHERE where place_id IN (123684,14854538,128578,14714287,311084,14778481,
       14844284,14794397,14496415,14389638,14445874,128021,14833047,14413001,
       110793,14713753,110580,14827990,123217,14778210);
```
so we can see if there's anything in the OSM data that looks fishy?

**Comment by boooch** [comment]
Sorry, already dropped database and started installation again. But if I did this request as 
UPDATE placex SET indexed_status = 0 WHERE place_id = 14854538;
UPDATE placex SET indexed_status = 0 WHERE place_id = 128578;
e.t.c 

was no errors. But after restarting setup.php --index 
next portion of ids  was in postgres error.
I think that some id ID in query linked with other IDS in thу same query but not sure.

**Comment by boooch** [comment]
Today process was broken on request: 
UPDATE placex SET indexed_status = 0 WHERE place_id IN (14517976,14787357,126877,14691953,123944,14647269,127620,14857562,125383,14707415,125992,14773977,627532,14658680,14693049,126897,14864744,123969,125404,14855891)

This is select you asket me :
nominatim=#  SELECT osm_id,osm_type
    FROM placex
    WHERE place_id IN (14517976,14787357,126877,14691953,123944,14647269,127620,14857562,125383,14707415,125992,14773977,627532,14658680,14693049,126897,14864744,123969,125404,14855891);
   osm_id   | osm_type
------------+----------
  424311738 | N
  424316022 | N
  424313903 | N
  424316493 | N
  424314335 | N
  424311662 | N
  424315932 | N
  424313751 | N
 3491215388 | N
     195273 | R
     171496 | R
    2093234 | R
     195266 | R
     192830 | R
     192798 | R
     195270 | R
  424311658 | N
     192763 | R
    1889339 | R
     296961 | R
(20 rows)

**Comment by boooch** [comment]
I`m using this dataset:
https://download.geofabrik.de/africa-latest.osm.pbf

**Comment by lonvia** [comment]
What version of Nominatim is this?

**Comment by lonvia** [comment]
#2046 has a fix for this, which you should be able to apply to the 3.5.2 version.

**Comment by boooch** [comment]
Confirm , now works fine!

**Comment by nashJG** [comment]
> Confirm , now works fine!

I got the save error when tried to indexing the osm planet data:
`psycopg2.errors.TriggeredDataChangeViolation: ERROR: tuple to be updated was already modified by an operation triggered by the current command`

The nominatim version is 3.5.2. How do you solve this problem?

**Comment by nashJG** [comment]
> #2046 has a fix for this, which you should be able to apply to the 3.5.2 version.

I have got the same indexing problem in 3.5.2version which downloaded from [https://nominatim.org/release/Nominatim-3.5.2.tar.bz2](url) .  The data is download from osm planet.
I also install nominatim following the instructions in [https://nominatim.org/release-docs/develop/](url). this nominatim is clone from git by the command `git clone --recursive git://github.com/openstreetmap/Nominatim.git`.   In this version, The indexing works fine.   But the search api doesn't work very well sometimes.  When I tried to search some address first time , it may take a long time, is this problem caused by my local postgresql server?

Now I have got the latest version from git following your advise. The init import program is running.

If I want to build a stable local nominatim service, which version should I choose?  How can I optimize  the local search service?
